File: src/main/java/aoc2019/Point.java
package aoc2019;

import java.util.stream.Stream;

public class Point {
    int x;
    int y;
    
    static final Point ZERO = of(0,0);

    public Point(int x, int y) {
        this.x = x;
        this.y = y;
    }

    static Point of(int x, int y) {
        return new Point(x,y);
    }
    
    int manhattan(Point a) {
        return Math.abs(a.x - x) + Math.abs(a.y - y);
    }
    
    public Point north() {
    	return Point.of(x,  y-1);
    }
    public Point south() {
    	return Point.of(x,  y+1);
    }
    public Point east() {
    	return Point.of(x+1,  y);
    }
    public Point west() {
    	return Point.of(x-1,  y);
    }
    
    public Stream<Point> neighbors() {
		return Stream.of(north(), south(), east(), west());
	}
    
    @Override
    public String toString() {
        return String.format("P[%d,%d]", x, y);
    }
    
    @Override
    public boolean equals(Object obj) {
        Point a = (Point)obj;
        return x==a.x && y==a.y;
    }
    
    @Override
    public int hashCode() {
        return x+3*y;
    }
}

File: src/main/go/aoc2019/Point.go
package main

// Point is a 2d coord
type Point struct {
	x, y int
}

func (p Point) manhattan() int {
	abs := func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}
	return abs(p.x) + abs(p.y)
}

func (p Point) mdist(p0 Point) int {
	abs := func(x int) int {
		if x < 0 {
			return -x
		}
		return x
	}
	return abs(p.x-p0.x) + abs(p.y-p0.y)
}

func (p Point) hash() int {
	return p.x + 10000*p.y
}

func (p Point) indexIn(w []Point) int {
	for i, pi := range w {
		if pi == p {
			return i
		}
	}
	return -1
}

func (p Point) north() Point {
	return Point{p.x, p.y - 1}
}

func (p Point) south() Point {
	return Point{p.x, p.y + 1}
}

func (p Point) east() Point {
	return Point{p.x + 1, p.y}
}

func (p Point) west() Point {
	return Point{p.x - 1, p.y}
}
